### PR TITLE
PHP 8.4 support

### PR DESCRIPTION
**Pull request by phpfui**
* Updated composer.json to support PHP 8.0 and higher only
* Moved to testing under GitHub Actions
* Ported to newer PHPUnit
* Added and corrected types for PHP 8.4 support

ToDo:
-----

Unit test need some work, but not sure what to do here as I am not familiar with the code.

### CODE REVIEW

File: src/Flow/Basic.php [hunk 1]
```diff
@@ -14,15 +14,13 @@ class Basic
     /**
      * @param  string                 $destination where to save file
      * @param  string|ConfigInterface $config
-     * @param  RequestInterface       $request     optional
-     * @return bool
      */
-    public static function save($destination, $config, RequestInterface $request = null)
+    public static function save(string $destination, $config, ?RequestInterface $request = null) : bool
```
Comment (AidasK) [inline src/Flow/Basic.php:18]: maybe type config as `string|ConfigInterface`

File: src/Flow/Config.php [hunk 2]
```diff
@@ -4,111 +4,86 @@
...
      * Set chunk identifier
-     *
-     * @param callable $callback
      */
-    public function setHashNameCallback($callback)
+    public function setHashNameCallback(callable | array $callback) : static
     {
         $this->config['hashNameCallback'] = $callback;
+
+		return $this;
```
Comment (AidasK) [inline src/Flow/Config.php:36]: not aligned?

File: src/Flow/Config.php [hunk 3]
```diff
@@ -4,111 +4,86 @@
...
      * Callback to pre-process chunk
-     *
-     * @param callable $callback
      */
-    public function setPreprocessCallback($callback)
+    public function setPreprocessCallback(callable | array $callback) : static
     {
         $this->config['preprocessCallback'] = $callback;
+
+		return $this;
```
Comment (AidasK) [inline src/Flow/Config.php:54]: align?

File: src/Flow/Config.php [hunk 4]
```diff
@@ -4,111 +4,86 @@
...
      * Delete chunks on save
-     *
-     * @param bool $delete
      */
-    public function setDeleteChunksOnSave($delete)
+    public function setDeleteChunksOnSave(bool $delete) : static
     {
         $this->config['deleteChunksOnSave'] = $delete;
+
+		return $this;
```
Comment (AidasK) [inline src/Flow/Config.php:72]: code style

File: src/Flow/Mongo/MongoConfig.php [hunk 5]
```diff
@@ -10,22 +10,16 @@
...
-    private $gridFs;
-
     /**
      * @param Bucket $gridFS storage of the upload (and chunks)
      */
-    function __construct(Bucket $gridFS)
+    function __construct(private Bucket $gridFS)
     {
         parent::__construct();
         $this->gridFs = $gridFS;
```
Comment (AidasK) [inline src/Flow/Mongo/MongoConfig.php:19]: remove this?

**Review by AidasK — CHANGES_REQUESTED** [review]

**Comment by AidasK** [comment]
Are these tests passing locally?

**Comment by phpfui** [comment]
Will fix these issues. Maybe I should set up phpcsfixer and phpstan too.

Tests fail for me locally, you will have to fix, as you know the code better. Not sure how the mocking works.

**Comment by phpfui** [comment]
I fixed the save() parameter type.  Also fixed the formatting with PHP-CS-Fixer.  Corrected a bunch of type errors with PHPStan (removed the MongoConfigInterface, as it is not really needed). I also fixed one of the PHPUnit errors.  The other two are caused by the mocking, but not exactly sure how to fix those, so you will need to work on them.

I am going to test this in my project.  Will push again if I find and fix any issues.  I can't test the MongoDB stuff, as I use SQL only.

**Comment by phpfui** [comment]
My local testing seems to work with no mods needed.  This should probably be released as v1.3 or v2.0

File: src/Flow/Mongo/MongoFile.php [hunk 6]
```diff
@@ -25,69 +24,53 @@ class MongoFile extends File
...
      */
-    public function chunkExists(int $index) : bool
+    public function chunkExists(int $index): bool
     {
-        return $this->config->getGridFs()->getChunksCollection()->findOne([
+        return null !== $this->config->getGridFs()->getChunksCollection()->findOne([
             'files_id' => $this->getGridFsFile()['_id'],
-            'n' => (intval($index) - 1)
-        ]) !== null;
+            'n' => ((int) $index - 1)
```
Comment (AidasK) [inline src/Flow/Mongo/MongoFile.php:43]: n was probably bool before?

**Comment by AidasK** [comment]
Released as 2.0, thank you for your contribution <3

**Comment by phpfui** [comment]
Thanks for the prompt attention. Seems PHP 8.4 compatibility is not a high priority for many maintainers.

**Pull request merged**